temporal-sdk/src/test/java/io/temporal/internal/replay/OutdatedDirectQueryReplayWorkflowRunTaskHandlerTest.java

```diff
@@ -0,0 +1,194 @@
+import io.temporal.workflow.shared.TestActivities;
+import io.temporal.workflow.shared.TestWorkflows;
+import java.time.Duration;
+import java.util.function.BiFunction;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+@Issue("https://github.com/temporalio/sdk-java/issues/1371")
+public class OutdatedDirectQueryReplayWorkflowRunTaskHandlerTest {
+  private static final String QUERY_RESULT = "queryIsDone";
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(WorkflowWithLocalActivityInASecondWorkflowTask.class)
+          .setActivityImplementations(new TestActivities.TestActivitiesImpl())
+          .build();
+
+  private WorkflowStateMachines stateMachines;
+
+  @Test
+  public void queryIsOutdated() {
+    TestWorkflows.TestWorkflowWithQuery noArgsWorkflow =
+        testWorkflowRule.newWorkflowStub(TestWorkflows.TestWorkflowWithQuery.class);
+    noArgsWorkflow.execute();
+    // check that a normal query works
+    assertEquals("A normal query should work", QUERY_RESULT, noArgsWorkflow.query());
+    WorkflowExecution workflowExecution = WorkflowStub.fromTyped(noArgsWorkflow).getExecution();
+
+    WorkflowExecutionHistory workflowExecutionHistory =
+        testWorkflowRule.getExecutionHistory(workflowExecution);
+
+    PollWorkflowTaskQueueResponseOrBuilder wft =
+        PollWorkflowTaskQueueResponse.newBuilder()
+            .setWorkflowExecution(workflowExecution)
+            .setPreviousStartedEventId(
+                WorkflowExecutionUtils.getEventsOfType(
+                        workflowExecutionHistory.getHistory(),
+                        EventType.EVENT_TYPE_WORKFLOW_TASK_STARTED)
+                    .get(0)
+                    .getEventId())
+            .setHistory(workflowExecutionHistory.getHistory())
+            .setQuery(WorkflowQuery.newBuilder().setQueryType("some-query").build());
+
+    ReplayWorkflowRunTaskHandler handler =
+        new ReplayWorkflowRunTaskHandler(
+            "UnitTest",
+            createReplayWorkflow(workflowExecutionHistory),
+            wft,
+            SingleWorkerOptions.newBuilder().build(),
+            new NoopScope(),
+            mock(BiFunction.class));
+
+    stateMachines = handler.getWorkflowStateMachines();
+    QueryResult queryResult =
+        handler.handleDirectQueryWorkflowTask(
+            wft, new FullHistoryIterator(workflowExecutionHistory.getEvents()));
+    assertEquals(
+        QUERY_RESULT,
+        DefaultDataConverter.STANDARD_INSTANCE.fromPayloads(
+            0, queryResult.getResponsePayloads(), String.class, String.class));
+  }
+
+  private ReplayWorkflow createReplayWorkflow(WorkflowExecutionHistory workflowExecutionHistory) {
+    History history = workflowExecutionHistory.getHistory();
+    ReplayWorkflow workflowMock = mock(ReplayWorkflow.class);
+    Mockito.doAnswer(
+            invocation -> {
+              stateMachines.newTimer(
```

Comment:
> In core what I do is manually construct a history that that workflow would produce when operating normally
We do it here very simply, by playing a normal workflow. 

> and then chunk it up into mock WFT poll responses from the server
We do it here the same way actually, we create a fake poll

This code is needed not to generate a history or something, but to bypass the whole layer that JavaSDK has around "workflow object" and "workflow class", because it's quite scattered. Core doesn't have this problem because it has this layer in lang. Basically, I work with "activations" directly here.

Reply:
I guess what I'm getting at is the state machine invocation itself seems a bit nitty-gritty compared to eventually seeing those commands in the WFT completion. It's not immediately clear to me why it has to operate down here instead of that level, but, if it's structurally too much stuff to make that happen I think this is still totally fine